File: core4/leetcode/Arrays/P108/main.cpp
#include <iostream>
#include <vector>

using namespace std;

// Node definition
class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};

// BinaryTree class
class BinaryTree {
public:
    Node* root;

    BinaryTree() {
        root = nullptr;
    }

    // Public function to start building the tree
    void BuildTree(vector<int>& nums) {
        root = buildSubtree(nums, 0, nums.size() - 1);
    }

    // Public inorder traversal
    void Inorder() {
        Inorder(root);
        cout << endl;
    }

private:
    // Helper recursive function to build BST
    Node* buildSubtree(vector<int>& nums, int start, int end) {
        if (start > end) {
            return nullptr; // base case: empty subarray
        }

        int mid = start + (end - start) / 2;
        Node* node = new Node(nums[mid]);

        // Recursively build left and right subtrees
        node->left  = buildSubtree(nums, start, mid - 1);
        node->right = buildSubtree(nums, mid + 1, end);

        return node;
    }

    // Helper recursive inorder traversal
    void Inorder(Node* node) {
        if (!node) return;
        Inorder(node->left);
        cout << node->data << " ";
        Inorder(node->right);
    }
};

int main() {
    vector<int> nums = {-10, -3, 0, 5, 9};

    BinaryTree tree;
    tree.BuildTree(nums);

    cout << "Inorder traversal of the BST: ";
    tree.Inorder(); // should print the sorted array
    return 0;
}

File: core4/leetcode/Arrays/P108/main.py
class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None
        
    
class Solution:
    def build_bst(self, nums, low, high):
        
        if low > high :
            return 
        
        mid = low + (high - low) // 2
        
        root = Node(nums[mid])
        root.left  = self.build_bst(nums, low, mid - 1)
        root.right = self.build_bst(nums, mid+1, high)
        
        return root
        
    def sorted_array_to_bst(self, nums):
        return self.build_bst(nums, 0, len(nums)-1)
        
        
    def display(self, root):
        if not root :
            return 
        
        self.display(root.left)
        print(root.data, end=" ")
        self.display(root.right)
        
        
        
# ---- Main ----
nums = [-10, -3, 0, 5, 9]

solution = Solution()
root = solution.sorted_array_to_bst(nums)

solution.display(root)
        
        

        
